agent/algorithm/docker/docker.go

```diff
@@ -0,0 +1,243 @@
+	return metadata.AppendToOutgoingContext(ctx, dockerRunCommandKey, runCommand)
+}
+
+func DockerRunCommandFromContext(ctx context.Context) string {
+	return metadata.ValueFromIncomingContext(ctx, dockerRunCommandKey)[0]
+}
+
+func New(logger *slog.Logger, eventsSvc events.Service, runCommand, algoFile string) algorithm.Algorithm {
+	d := &docker{
+		algoFile: algoFile,
+		logger:   logger,
+		stderr:   &algorithm.Stderr{Logger: logger, EventSvc: eventsSvc},
+		stdout:   &algorithm.Stdout{Logger: logger},
+	}
+
+	if runCommand == "" {
+		d.runCommand = DockerRunCommand
+	} else {
+		d.runCommand = runCommand
+	}
+
+	return d
+}
+
+func (d *docker) Run() ([]byte, error) {
+	ctx := context.Background()
+
+	// Create a new Docker client.
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return []byte{}, fmt.Errorf("could not create a new Docker client: %v", err)
+	}
+
+	// Open the Docker image tar file.
+	imageFile, err := os.Open(d.algoFile)
+	if err != nil {
+		return []byte{}, fmt.Errorf("could not open Docker image: %v", err)
+	}
+	defer imageFile.Close()
+
+	// Load the Docker image from the tar file.
+	resp, err := cli.ImageLoad(ctx, imageFile, true)
+	if err != nil {
+		return []byte{}, fmt.Errorf("could not load Docker image from file: %v", err)
+	}
+	defer resp.Body.Close()
+
+	// List the loaded images to get the image ID.
+	images, err := cli.ImageList(ctx, image.ListOptions{})
+	if err != nil {
+		return []byte{}, fmt.Errorf("could not get the Docker image list: %v", err)
+	}
+
+	var imageID string = ""
+	var dockerImageName string = ""
+	for _, image := range images {
+		for _, tag := range image.RepoTags {
+			imageID = image.ID
+			dockerImageName = tag
+			break
+		}
+	}
+
+	if imageID == "" {
+		return []byte{}, fmt.Errorf("could not find image ID")
+	}
+
+	dockerCommand := strings.Fields(d.runCommand)
+
+	// Create and start the container.
+	respContainer, err := cli.ContainerCreate(ctx, &container.Config{
+		Image: dockerImageName,
+		Cmd:   dockerCommand,
+	}, &container.HostConfig{
+		Mounts: []mount.Mount{
+			{
+				Type:   mount.TypeBind,
+				Source: algorithm.DatasetDirectory,
+				Target: algorithm.DatasetDirectory,
+			},
+		},
+	}, nil, nil, containerName)
+	if err != nil {
+		return []byte{}, fmt.Errorf("could not create a Docker container: %v", err)
+	}
+
+	if err := cli.ContainerStart(ctx, respContainer.ID, container.StartOptions{}); err != nil {
+		return []byte{}, fmt.Errorf("could not start a Docker container: %v", err)
+	}
+
+	statusCh, errCh := cli.ContainerWait(ctx, respContainer.ID, container.WaitConditionNotRunning)
+	select {
+	case err := <-errCh:
+		if err != nil {
+			return []byte{}, fmt.Errorf("could not wait for a Docker container: %v", err)
+		}
+	case <-statusCh:
+	}
+
+	agentResultOutput := path.Join(algorithm.AlgorithmResultOutputDir, algorithm.AlgorithmResultOutputFile)
```

Comment:
there could be multiple files. results dir is zipped by agent

Reply:
The code is now changed to reflect this change.